Simplenote/SPAppDelegate.m

```diff
@@ -812,11 +815,72 @@ - (BOOL)application:(UIApplication *)application handleOpenURL:(NSURL *)url
         [_simperium save];
         
         [self presentNote:newNote];
+    } else if ([[url host] isEqualToString:@"auth"]) {
+        NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:url
+                                                    resolvingAgainstBaseURL:NO];
+        NSArray *queryItems = urlComponents.queryItems;
+        NSString *state = [self valueForKey:@"state" fromQueryItems:queryItems];
+        NSString *user = [self valueForKey:@"user" fromQueryItems:queryItems];
+        NSString *token = [self valueForKey:@"token" fromQueryItems:queryItems];
+        NSString *isNew = [self valueForKey:@"new" fromQueryItems:queryItems];
+        NSString *wpcomToken = [self valueForKey:@"wp_token" fromQueryItems:queryItems];
+        
+        if (state == nil || user == nil || token == nil) {
+            [[NSNotificationCenter defaultCenter] postNotificationName:kSignInErrorNotificationName
+                                                                object:nil];
+            return false;
+        }
+        
+        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+        NSString *storedState = [defaults stringForKey:@"SPAuthSessionKey"];
+        if (![state isEqualToString: storedState]) {
+            // States don't match!
+            [[NSNotificationCenter defaultCenter] postNotificationName:kSignInErrorNotificationName
+                                                                object:nil];
+            return false;
+        }
+        
+        if (self.simperium.user.authenticated) {
+            // We're already signed in
+            [[NSNotificationCenter defaultCenter] postNotificationName:kSignInErrorNotificationName
+                                                                object:nil];
+            return false;
+        }
+        
+        [defaults removeObjectForKey:@"SPAuthSessionKey"];
+        [defaults setObject:user forKey:@"SPUsername"];
+        NSError *error = nil;
+        BOOL success = [SPKeychain setPassword:token forService:[SPCredentials simperiumAppID] account:user error:&error];
+        if (success == NO) {
+            [[NSNotificationCenter defaultCenter] postNotificationName:kSignInErrorNotificationName
+                                                                object:nil];
+            return false;
+        }
+        
+        // Store wpcom token for future integrations
+        if (wpcomToken != nil) {
+            [SPKeychain setPassword:token forService:kSimplenoteWPServiceName account:user error:&error];
+        }
+        
+        self.simperium.user = [[SPUser alloc] initWithEmail:user token:token];
+        [self.navigationController dismissViewControllerAnimated:YES completion:nil];
+        [self.simperium authenticationDidSucceedForUsername:user token:token];
+        
+        if ([isNew isEqualToString:@"true"]) {
+            [self createWelcomeNoteAfterDelay];
+        }
```

Comment:
Suggestion!!

Do you think it would be much of a trouble to move this whole new block of code to a new class, for encapsulation purposes?. Something like...

```
@interface WordPressAuthenticationHandler
+ (BOOL)isWordPressAuthenticationUrl:(NSURL*)url;
+ (SPUser *)extractSimplenoteUserFromUrl:(NSURL*)url;
@end
```

Reply:
(It'd be neat if we could, probably, have another class, App-Level, to encapsulate all of the authentication as well).

```
- (BOOL)application:(UIApplication *)application handleOpenURL:(NSURL *)url
{    
     //  ..
     else if ([WordPressAuthenticationHandler supportsUrl:url]) {
         SPUser *user = [WordPressAuthenticationHandler extractSimplenoteUserFromUrl:url];
         if (user == nil) {
             // error handling
            return;
         }

         [[SimplenoteAuthenticator shared] updateSimperiumUser:user];
     }
}
```

(ROUGH, very ROUGH idea), but the goal would be to move a little bit of the logic elsewhere, so that we begin cleaning up a little bit the AppDelegate. Lemme know what you think!!

Reply:
Nice idea, I added a new `WPAuthHandler` class in 3495c30. I also added support for a better error message if the user hasn't activated their wp.com account yet.